### ISSUE TITLE
Add a way of getting group counts to `h`

### ISSUE THREAD

**Issue by jon-betts** [hypothesis/h#7857]
We need to be able to retrieve counts of annotations in groups for multiple users at a time.

The idea here is to return the counts of annotations by user in each group, to allow LMS to then sort out how to combine those numbers based on who is, or isn't, a teacher. This allows us to return one set of results which could potentially be used to provide tailored results for each teacher.

## Some features

 * Can accept multiple viewers or groups to look for **(Required)**
 * Can accept a date range or recency (like "last 24 hours") **(Required)**
 * Can return a large number of results **(Required)**
 * Must hide NIPSA'd users annotations **(Required)** 
 * Can expand gracefully to return more information (like example annotation bodies or other anno info) _(Nice to have)_
 * Similar to the [existing formats](https://h.readthedocs.io/en/latest/api-reference/v2/#tag/annotations/paths/~1search/get) used _(Nice to have)_

## Tickets

 * https://github.com/hypothesis/h/pull/7858 - Spike
 * [x] Agree a general design for the API (outlines below)
 * [x] https://github.com/hypothesis/h/pull/7860
 * [x] https://github.com/hypothesis/h/pull/7859
 * [x] https://github.com/hypothesis/h/pull/7863
 * [x] https://github.com/hypothesis/h/pull/7862
 * [x] https://github.com/hypothesis/h/pull/7865
 * [x] https://github.com/hypothesis/lms/pull/5132

###  Permissions

The current bulk API has a permission (`Permission.API.BULK_ACTION`) which arguably fits this situation quite well. It was intended for the bulk API end-point alone, but the semantics seem good and it's already setup to allow connections from LMS only.

### Query params

<details><summary>Existing conventions around query params</summary>

https://h.readthedocs.io/en/latest/api-reference/v2/#tag/annotations/paths/~1annotations/post

These are all `GET`:

> search_after  
> 
> ISO 8601 string / milliseconds since epoch
> search_after=2019-01-03T19%3A46%3A09.334539%2B00%3A00
> 
> sort
> 
> Default: "updated"
> Enum: "created" "updated" "group" "id" "user"
> 
> user
> 
> Example: user=acct:username@hypothes.is
> Limit the results to annotations made by the specified user.
> 
> group	
> 
> Example: group=8JmD3iz1
> Limit the results to annotations made in the specified group (by group ID).
> 
> This can be specified multiple times to retrieve multiple groups.
</details>

We need a term for differentiating between the user who is viewing the annotation vs. the author of the annotation. 

In most calls the author is referred to as the `user` the viewer is assumed from the logged in user. In the case of bulk actions there is no logged in user. In the case of the Bulk API this was solved by having `user.effective` (JSON) or `effective_user` (code). This is the user which LMS is acting on behalf of, which is largely the same as a viewer.

### Body format

<details><summary>Annotation JSON format</summary>

```json
{
    "id": "string",
    "created": "2023-02-21T14:18:35Z",
    "updated": "2023-02-21T14:18:35Z",
    "user": "acct:felicity_nunsun@hypothes.is",
    "uri": "http://example.com",
    "text": "string",
    "tags": ["string"],
    "group": "string",
    "permissions": { },
    "target": {
        "source": "http://example.com",
        "selector": []
    },
    "links": { },
    "hidden": true,
    "flagged": true,
    "references": ["string"],
    "user_info": {
        "display_name": "Felicity Nunsun"
    }
}
```

</details>

In the Bulk API we chose to use _some_ JSON:API trappings as an aid to design. In this case it's likely to cause a lot of bloat, so we might want to avoid it.

The Bulk API also makes use of [NDJSON](http://dataprotocols.org/ndjson/), which is very well optimised for mass transport of individual documents. This allows large bodies (hundreds of megabytes or more) to be streamed from a producer to a consumer without either having to keep the results in memory at any time. This could be a good fit for this application.

<details><summary>Creating streaming NDJSON in Pyramid</summary>

```python
# When we get an iterator we must force the first return value to be
# created to be sure input validation has occurred. Otherwise we might
# raise errors outside of the view when called

try:
    results = chain(  # pylint: disable=redefined-variable-type
        [next(results)], results
    )
except StopIteration:
    results = []

return Response(
    app_iter=((json.dumps(result) + "\n").encode("utf-8") for result in results),
    status=200,
    content_type="application/x-ndjson",
)
```

Parsing is similarly easy:

```python
with requests.Session().get(url, stream=True) as response:
    for line in response.iter_lines():
        data = json.loads(line)
```
</details>

### Method (GET / POST)

Using a GET request is more similar to existing practice, and is more convenient for certain callers, but POST allows us to use JSON as a payload, which means more straight forward nesting of arguments etc.

We only have one caller (LMS), and any future callers are likely not going to be using this end-point directly and will have different semantics. For example, anything from the client _will_ be from the point of view of the caller being the effective user / viewer, and we will want to enforce this.

Another concern is length. If we pass full user ids from LMS like: `acct:b93e505e9dcbec146f966a4161741a@lms.ca.hypothes.is` we have 55 chars per username (including a comma). This means assuming there were no other content in get params the total number we'd get before we start hitting likely limits is:

 * 2Kb limit - 37 user ids (66 usernames)
 * 8Kb limit - 148 user ids (264 usernames)

I'm not sure at which point NGINX is going to throw it's dummy out (some things seem to suggest 8k). We could use some tricks to get around this (like passing the username only and the authority separately). But if we go with POST, the default is 1Mb which allows for 19k user ids (34k usernames). For reference we have ~15k teachers in the US. So this method could conceivably do them all in one call.

<details><summary>I can't find any practical limit to the size of in IN query in Postgres...</summary>

```python
import pytest

from lms.models import User


class TestWoo:
    @pytest.mark.parametrize("num_ids", (
        10,
        1000,
        5000,
        10000,
        100000,
        1000000,
    ))
    def test_it(self, db_session, num_ids):
        wat = db_session.query(User).filter(
            User.user_id.in_(list(f"name_{i}" for i in range(num_ids)))
        ).count()

        print(num_ids, wat)
```
It gets slower with a million, but never stops.

</details>

**Preference**: POST

### Params

There are a number of already defined parameters for this:

 * `group` - Get param, specify multiple times
 * `search_after` - Timestamp, applies to 
 * `user` - Author of annotation
 
But these aren't quite what we want (group is the wrong id, user is the wrong user) or a bit vague (search after what?).

<details><summary>Query to get courses / teacher</summary>

```sql
SELECT 
    AVG(COUNT)
FROM (
    SELECT
        COUNT(1) AS count
    FROM report.group_roles
    JOIN report.groups ON 
        groups.id = group_roles.group_id
    WHERE 
        role = 'teacher'
        AND group_type = 'course'
    GROUP BY
        user_id
) AS data
```
</details>

According to the above, the average number of courses per teacher in the US is 3.25. This means if we had 100 users to retrieve, we could either pass 100 user ids, or 325 group ids. For this reason we probably want to pass the user ids. User ids are quite long (55 chars), but if we pass them as usernames only (30 chars) they are shorter than authority provided ids as well (40 chars).

This means we need to tackle the viewer question. We also need a date range for updated / created. Something like this might do:

```json
{
    "filter": {
        "user": {"effective": ["acct:...", "acct:..."]},
        "updated": {"gt": "...", "lte": "..."}
    },
    "fields": ["group", "user"]
}
```

**Preference**: JSON as above

### Body Payload (raw counts dict / list of dicts)

The most compact way to return the data we need is to return raw counts for groups and users:

```json
{
   "groupid": {
        "userid": 234,
        "userid2": 234,
   }
}
```
But this doesn't give us much scope for returning more information in the future. A more flexible format would be to return one row per annotation without much more detail:

```json
[
    {"group": "groupid", "user": "userid"},
    {"group": "groupid", "user": "userid"},
    {"group": "groupid", "user": "userid"},
    {"group": "groupid", "user": "userid2"},
]
```

Unfortunately we don't have the group ids or public ids in LMS, we only have the authority provided ids. In a horrible turn of events we refer to the `authority_provided_id` as `group_id` in the [group API](https://h.readthedocs.io/en/latest/api-reference/v2/#tag/groups/paths/~1groups/post), which also calls the `groupid` just `id` which we call `group` in the annotation context.

The best idea I've got so far is to sort of ignore the existing annotation format for these fields, and instead use the same format you would get if you call the `/group` and `/user` parts of the API:

```json
{
    "user": {
        "username": "928ejiwuenrinwoeuirwye",
    },
    "group": {
        "groupid": "eue234irwyo342928ejiwuenrinw3456"
    }
}
```

The `groupid` here is still `authority_provided_id`, but at least it's the same flavour of confusing as elsewhere. If we wanted to get very fancy, we could change "user" to "author" here.

### Body payload wrapper (None / JSON:API / NDJSON / CSV)

These different options are somewhat interchangeable, and you could mix and match certain combinations.

#### No wrapper

Assuming we choose a plain list or dict above, we can return it verbatim. 

#### JSON:API

Assuming we have aspirations to be JSON:API flavoured, we might want to nest in whatever we do one layer with a "data" key. This is kind of a good idea in this scenario anyway (even if we don't follow any other parts of the format) as it easily allows for other types of metadata if you want some.

#### NDJSON

NDJSON is simple as dirt and is the least decorated of the formats. It takes a little more effort to generate and parse, but is the only format that supports streaming arbitrarily large amounts of data.

The consumer could conceivably start doing work based on the data it sees before the producer has finished executing the query.

_Question_: How does GZIP change this scenario? Our return format is likely to have lots of repeated data and keys. I'm intuition is that streaming would make GZIP more difficult / disable it? Could this be a false economy for certain sizes of report? I still lean to memory being more important than total bandwidth. **Update** It seems like this might have implications for GZIP, but it can handle streaming data: https://stackoverflow.com/questions/19700503/how-does-gzip-compression-rate-change-when-streaming-data

#### CSV (Dark horse suggestion)

CSV has a pretty great feature of being streamable like NDJSON and also not including any repeated keys. So it's actually quite compact.

However serialising a complicated nested structure to CSV might be kind of a pain. This is especially true if it includes any variable length lists. This would be very horrible to work out. For any static-ish set of nested dict keys, this would actually be fine.

#### Notes on size

There were about 35k annotations in the last 24 hours.

| Format | Streamable | Minimal | Maximal | Read Time (Minimal) |
|--------|------------|---------|---------|---------------------|
| JSON   | No         | ~5Mb    | ~68Mb   | 60-200ms            |
| NDJSON | Yes        | ~5Mb    | ~68Mb   | 60-200ms            |
| CSV    | Yes        | ~2.7Mb  | ~50Mb   | 50ms                |

For this exact number and size of data I'd say it's a wash. _However_ this is only true as we are retrieving basically no fields for the data. A commonly requested thing is for unis to get a dump of all annotations made. In this case they would want a lot more fields and the size would balloon. This is also true if we wanted to include annotation text or something else.

Using an example full annotation body and regenerating I get about 68Mb for 35k (this is totally arbitrary as there is no limit to the size of an annotation, the could be 1Mb each). The largest uni has 577,929 annotations at the moment which would put it a touch over 1Gb.

CSV is actually kind of a win for some of this, coming in at 2.7Mb / 50Mb for the minimal / maximal situation.

**Preference:** NDJSON / then whatever we decide on in a "data" key

### Request

```
POST /api/bulk/annotation HTTP/1.1
Content-Type: application/json
Accept: application/vnd.hypothesis.v1+x-ndjson

{
    "filter": {
         "audience": {
             "username": [..., ..., <POTENTIALLY VERY LARGE LIST HERE> ...]
         },
         "updated": {"gt": "2023-02-21T18:32:27", "lte": "2023-02-21T18:32:27"},
         "limit": 100000
     },
     "fields": ["author.username", "group.authority_provided_id"]
}
```

_Note_: We don't need to get the authority provided to us explicitly. I think we can derive this from the authentication information.

_Note_: I'm not sure about effective user. This is who we are doing it on behalf of, which does seem mostly right? Maybe "viewer" or "reader" is clearer?

### Response

```
HTTP/1.1 200 OK
Hypothesis-Media-Type: application/vnd.hypothesis.v1+x-ndjson
Content-Type: application/x-ndjson

{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
{"author": {"username": "928ejiwuenrinwoeuirwye"}, "group": {"authority_provided_id": "eue234irwyo342928ejiwuenrinw3456"}}
...
```

_Note:_ If we want to hew closer to existing practice we could use "user" instead of "author" here

### Implementation notes

 * Hide all NIPSA'd users full stop
 * We probably want some kind of limit, even if it's very high
 * If we want pagination, then limit offset seems like a simple way to go
    * If we want a limit offset, then we need a sort order
    * Do we want pagination? Would be nice if we could avoid it
    * You can ask for fewer users at once
    * What if we had a return type of `{"type": "and_much_much_more"}`?
    * We could use this to return 50 records and then have the caller say 50+ for example
 * Sorting might be a good idea for streaming and consumption:
    * If similar data follows one after the other, then it will compress better
    * If we group by group, then when the group changes from one to another, you can start calculating numbers right away
